### PR TITLE
Server unittests

### PR DESCRIPTION
**Pull request by veloman-yunkan**
The server code was missing real unit-tests. The pre-existing unit test `test/kiwixserve.cpp` tests kiwix-lib indirectly by running the `kiwix-serve` executable which may be built with another version of the library (and involves a different repository `kiwix-tools`).

This PR adds the cpp-httplib C++ HTTP/HTTPS library (https://github.com/yhirose/cpp-httplib, v0.5.11). The new file `test/httplib.h` was copied from https://raw.githubusercontent.com/yhirose/cpp-httplib/v0.5.11/httplib.h. Only the client part of the library is needed by the new unit-test, however it is added in its entirety.

Included with this PR is an improvement in the meson setup that generates `kiwixlib-resources.h` and `kiwixlib-resources.cpp` only as needed (instead of considering them always stale and recompiling on every build). Without that change the edit-build-run cycle during a TDD-style development (supported by fast unit-tests) is not quick enough.

This unit test was created in the course of the work on kiwix/kiwix-tools#140, a significant part of which is now complete under #340. However that PR started growing too large, and I decided to break it into several smaller ones. This PR is the first of them.

In order to make the unit-tests more fail safe, the server was enhanced with the capability of starting on a randomly selected free port. However that change failed to build under **Linux (native_dyn)** & **Linux (win32_dyn)**, so it was reverted.

### CODE REVIEW

**Comment by veloman-yunkan** [comment]
> I'm a bit puzzled about having a copy of httplib in our source tree.
> 
> In one hand it simplify a lot our build system. But on the other we used to not include dependencies in our source-tree. We had gtest included and we've removed it.
> 
> I would like to say you to include the files because it is far less work but in the same time it is not the "right" thing to do.
> 
> (If we go this way, this also open a way to include other small library (mustache). For the best and the worst)
> 
> Independently of if we include libhttp or not, if we start to use it in the test, we may also consider to use it in the library to replace libcurl and libmicrohttpd.

cpp-httplib was added to this repository only as a shortcut. I fully support converting it to a base dependency before using it in production code.

File: static/meson.build [hunk 1]
```diff
@@ -1,4 +1,17 @@
...
+if cat.found()
+    resource_files = run_command('cat',
+                        meson.current_source_dir(),
+                        'resources_list.txt'
+                     ).stdout().strip().split('\n')
+    always_build_lib_resources = false
+else
+    resource_files = 'resources_list.txt'
+    always_build_lib_resources = true
+endif
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply static/meson.build:13]: @mgautierfr Should I create a new script file, or providing the script text via `python -c "..."` is acceptable?

**Comment by kelson42** [comment]
> cpp-httplib was added to this repository only as a shortcut. I fully support converting it to a base dependency before using it in production code.

We should secure as well that this is properly managed within kiwix-build.

File: static/meson.build (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgautierfr) [reply static/meson.build:13]: For a small script like that `python -c` is good.

[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply static/meson.build]: Fixed in a separate commit. Will merge it into the original commit after the CI run completes.

**Comment by veloman-yunkan** [comment]
@mgautierfr While preparing this PR for merging and squashing certain commits, should I preserve the unsuccessful change [Server can be started on a random free port](https://github.com/kiwix/kiwix-lib/pull/344/commits/99fd6db900c2e1f0e255387cdd63b778ceae6cae) together with its [reversal](https://github.com/kiwix/kiwix-lib/pull/344/commits/2ca19a6f98280deabb7612ef00e5763b3041067f)?

**Comment by veloman-yunkan** [comment]
> @mgautierfr While preparing this PR for merging and squashing certain commits, should I preserve the unsuccessful change [Server can be started on a random free port](https://github.com/kiwix/kiwix-lib/pull/344/commits/99fd6db900c2e1f0e255387cdd63b778ceae6cae) together with its [reversal](https://github.com/kiwix/kiwix-lib/pull/344/commits/2ca19a6f98280deabb7612ef00e5763b3041067f)?

@mgautierfr I wonder if you missed this question

**Comment by mgautierfr** [comment]
> should I preserve the unsuccessful change Server can be started on a random free port together with its reversal?

Yes. Try to keep a history of the PR clean (keep the logical steps).
However, it is good to not lost the code itself. Maybe you can push the commit on a different branch and open a issue pointing to it and explaining the situation.

**Comment by veloman-yunkan** [comment]
@mgautierfr Your answer is a little ambiguous. So should I keep those two commits on the `server_unittests` branch or remove them?

**Comment by kelson42** [comment]
> @mgautierfr Your answer is a little ambiguous. So should I keep those two commits on the `server_unittests` branch or remove them?

@veloman-yunkan I would say, try to keep it here. There is not only one manner to make things here, just try to keep the revision log understandable, with not too big commits, nice messages and here it might be nice to keep trace of this attempt.

**Comment by veloman-yunkan** [comment]
@kelson42 Done. IMO, this PR is ready for merging.

**Comment by kelson42** [comment]
@mgautierfr Could you please review again this PR? I have some relatively serious concern to put httplib code directly in our git repo (even temporary).
@veloman-yunkan Thank you very much for the last commits.

**Comment by mgautierfr** [comment]
> Your answer is a little ambiguous. So should I keep those two commits on the server_unittests branch or remove them?

My bad. I have read "should I squash them ?". You should squash them (and if one is the opposite of the other, remove them).
However, the code itself is good (even if it doesn't work because of libmicrohttpd version), it would be nice to keep it in another PR.

> I have some relatively serious concern to put httplib code directly in our git repo (even temporary).

I have even more concern about this when I think about this.
This add a dependency to httplib only for the tests. And I would prefer to avoid to add a dependency to a library only used in tests.
This is especially true as we already have libcurl as client library. We could use it instead of adding another library.

(I'm open to the idea to replace libmicrohttpd and libcurl by httplib, but that is another discussion)

**Comment by kelson42** [comment]
@mgautierfr This is not only for test I believe. @veloman-yunkan could confirm, but this is probably going to be used for next PRs to a proper mgmt of byte ranges requests.

**Comment by veloman-yunkan** [comment]
@mgautierfr @kelson42 As discussed in Slack, now the single-file version of cpp-httplib is being used (the version is mentioned both in the PR and the commit introducing it).

**Comment by kelson42** [comment]
@veloman-yunkan Thank you very much. I don’t see why we sould not merge now.

**Pull request merged**